src/transport/FabricTable.cpp

```diff
@@ -517,6 +474,34 @@ CHIP_ERROR FabricTable::LoadFromStorage(FabricIndex id)
     return err;
 }
 
+CHIP_ERROR FabricTable::AddNewFabric(FabricInfo & newFabric, FabricIndex & assignedIndex)
```

Comment:
I really wish we were not so deeply embedding the concept of FabricIndex into all corners of the stack.

I realize it is a convenient key to find operational identities.  But a pointer to an OpCreds object would do just as well for internal use.  FabricIndex is only necessary as a concept for the externally facing Node Operational Credentials Cluster API.  And not all nodes need to expose that cluster.  For instance, commissioners that bootstrap themselves onto fabrics by ecosystem-specific means presumably wouldn't.

By doing things this way, we are forcing all implementations to use the `OperationalCredentialSet` object, which defines the maximum number of supported fabrics at compile time.  That's extremely inconvenient for some use cases.  If a node has memory to accommodate an unknown but large number of fabrics, the sdk should not be a constraint.  But instead, we're embedding something very restrictive derived from the Node Operational Credentials Cluster spec.

Reply:
The AddNOC command requires cluster server to return the fabric index to the commissioner. So, we have to return the assigned fabric index here.

I do agree that we need to overhaul the Fabric Table/Info, and OpCredSet data structures. This PR is mainly trying to fix the assignment of node ID and fabric ID (i.e. get it from opCreds, rather than inferring from the received messages).

Reply:
To add some spec-centric background, what I am pointing out is that the spec's description under "4.9.2.1. Session Context" may be incorrect:

> 9. Peer Fabric Index: Records the Fabric Index used by the session’s peer to refer to the session’s Fabric and used to identify which fabric-scoped entries belong to the session’s common Fabric when such indices appear in messages.
◦ This field SHALL contain the "no Fabric" value of 0 when the SessionType is PASE and successful invocation of the AddNOC command has not yet occurred during commissioning.

As far as I can tell, only commissioners would have this information.  Non-administrator entities establishing a CASE session shouldn't need this information and I don't see how they would acquire it.

Reply:
Yes, that's correct. This index is only used for managing the fabrics on a device. Intent is not to use it for setting up CASE sessions (it might be used that way right now).